core/src/main/scala/com/phaller/rasync/Cell.scala

```diff
@@ -479,85 +635,117 @@ private class CellImpl[K <: Key[V], V](pool: HandlerPool, val key: K, updater: U
   @tailrec
   private def tryCompleteAndGetState(v: Try[V]): AnyRef = {
     state.get() match {
-      case current: State[_, _] =>
-        val currentState = current.asInstanceOf[State[K, V]]
-        val newVal = Success(tryJoin(currentState.res, v.get))
-        if (state.compareAndSet(current, newVal))
-          (currentState, newVal)
+      case current: IntermediateState[_, _] =>
+        val currentState = current.asInstanceOf[IntermediateState[K, V]]
+        val newVal = tryJoin(currentState.res, v.get)
+
+        // Copy completeDependentCells/nextDependentCells from IntermediateState to FinalState.
+        val newCompleteDependentCells: TrieMap[Cell[K, V], Unit] = new TrieMap()
+        val newNextDependentCells: TrieMap[Cell[K, V], Unit] = new TrieMap()
+        currentState.completeDependentCells.foreach(c => newCompleteDependentCells.put(c, ()))
+        currentState.nextDependentCells.foreach(c => newNextDependentCells.put(c._1, ()))
+        val newState = new FinalState[K, V](
+          Success(newVal),
+          newCompleteDependentCells,
+          newNextDependentCells)
+
+        if (state.compareAndSet(current, newState))
+          (currentState, newState)
         else
           tryCompleteAndGetState(v)
 
-      case finalRes: Try[_] => finalRes
+      case finalRes: FinalState[_, _] => finalRes
     }
   }
 
-  override def tryComplete(value: Try[V]): Boolean = {
+  override def tryComplete(value: Try[V], dontCall: Option[Seq[Cell[K, V]]]): Boolean = {
     val resolved: Try[V] = resolveTry(value)
 
     // the only call to `tryCompleteAndGetState`
     val res = tryCompleteAndGetState(resolved) match {
-      case finalRes: Try[_] => // was already complete
-        val finalResult = finalRes.asInstanceOf[Try[V]].get
-        val newVal = value.map(tryJoin(finalResult, _))
-        val res = finalRes == newVal
-        res
-
-      case (pre: State[K, V], newVal: Try[V]) =>
-        val nextCallbacks = pre.nextCallbacks
-        val completeCallbacks = pre.completeCallbacks
-
-        if (nextCallbacks.nonEmpty)
-          nextCallbacks.values.foreach { callbacks =>
-            callbacks.foreach(callback => callback.execute())
-          }
-        if (completeCallbacks.nonEmpty)
-          completeCallbacks.values.foreach { callbacks =>
-            callbacks.foreach(callback => callback.execute())
-          }
+      case _: FinalState[K, V] => // was already complete
+        true // As decided by phaller,  we ignore all updates after freeze and do not throw exceptions
+
+      case (pre: IntermediateState[K, V], finalValue: FinalState[K, V]) =>
+        // Inform dependent cells about the update.
+        val dependentCells = pre.nextDependentCells.keys ++ pre.completeDependentCells
+        dontCall match {
+          case Some(cells) =>
+            dependentCells.foreach(c => if (!cells.contains(c)) c.updateDeps(this))
```

Comment:
This might evaluate `!cells.contains(c)` quite often. Perhaps first compute `dependentCells - cells` and update only those?